File: Backend/src/Iit/Iit.Fibertest.Graph/Algorithms/Gis/GisLabCalculator.cs
using GMap.NET;

namespace Iit.Fibertest.Graph
{
    public static class GisLabCalculator
    {
        const int EarthRadius = 6372795;

        public static double GetDistanceBetweenPointLatLng(PointLatLng p1, PointLatLng p2)
        {
            var q1 = new CoorsInRad(p1);
            var q2 = new CoorsInRad(p2);
            return GetDistanceBetweenPoints(q1, q2);
        }

        public static PointLatLng GetPointAsPartOfSegment(PointLatLng p1, PointLatLng p2, double part)
        {
            var q1 = new CoorsInRad(p1);
            var q2 = new CoorsInRad(p2);

            // обратная геодезическая задача
            var decCoors = SphereToDecart(q2);
            decCoors = RotateAround(decCoors, q1.Lng, DecartAxis.Z);
            decCoors = RotateAround(decCoors, Math.PI / 2 - q1.Lat, DecartAxis.Y);
            CoorsInRad q = DecartToSphere(decCoors);
            double azimuth = Math.PI - q.Lng;
            var distanceOnSphere = Math.PI / 2 - q.Lat;

            // часть отрезка
            var distanceOnSphere2 = distanceOnSphere * part;

            // прямая геодезическая задача
            var sph = new CoorsInRad(Math.PI / 2 - distanceOnSphere2, Math.PI - azimuth);
            var decartCoors = SphereToDecart(sph);
            decartCoors = RotateAround(decartCoors, q1.Lat - Math.PI / 2, DecartAxis.Y);
            decartCoors = RotateAround(decartCoors, -q1.Lng, DecartAxis.Z);
            var coorsInRad = DecartToSphere(decartCoors);
            return new PointLatLng(coorsInRad.Lat / Math.PI * 180, coorsInRad.Lng / Math.PI * 180);
        }

        private static double GetDistanceBetweenPoints(CoorsInRad q1, CoorsInRad q2)
        {
            var decCoors = SphereToDecart(q2);
            decCoors = RotateAround(decCoors, q1.Lng, DecartAxis.Z);
            decCoors = RotateAround(decCoors, Math.PI / 2 - q1.Lat, DecartAxis.Y);
            var q = DecartToSphere(decCoors);
            var distanceOnSphere = Math.PI / 2 - q.Lat;
            var distance = distanceOnSphere * EarthRadius;
            return distance;
        }


        private static DecartCoors SphereToDecart(CoorsInRad sphere)
        {
            double p = Math.Cos(sphere.Lat);
            var result = new DecartCoors
            {
                Z = Math.Sin(sphere.Lat),
                Y = p * Math.Sin(sphere.Lng),
                X = p * Math.Cos(sphere.Lng)
            };

            return result;
        }

        private static CoorsInRad DecartToSphere(DecartCoors coors)
        {
            var hypot = Math.Sqrt(coors.X * coors.X + coors.Y * coors.Y);
            return new CoorsInRad
            {
                Lng = Math.Atan2(coors.Y, coors.X),
                Lat = Math.Atan2(coors.Z, hypot)
            };
        }

        private static DecartCoors RotateAround(DecartCoors coors, double alpha, DecartAxis axes)
        {
            switch (axes)
            {
                case DecartAxis.X:
                    return new DecartCoors()
                    {
                        X = coors.X,
                        Y = coors.Y * Math.Cos(alpha) + coors.Z * Math.Sin(alpha),
                        Z = -coors.Y * Math.Sin(alpha) + coors.Z * Math.Cos(alpha),
                    };
                case DecartAxis.Y:
                    return new DecartCoors()
                    {
                        X = -coors.Z * Math.Sin(alpha) + coors.X * Math.Cos(alpha),
                        Y = coors.Y,
                        Z = coors.Z * Math.Cos(alpha) + coors.X * Math.Sin(alpha),
                    };
                // case DecartAxis.Z:
                default:
                    return new DecartCoors()
                    {
                        X = coors.X * Math.Cos(alpha) + coors.Y * Math.Sin(alpha),
                        Y = -coors.X * Math.Sin(alpha) + coors.Y * Math.Cos(alpha),
                        Z = coors.Z,
                    };
            }
        }

        public static PointLatLng GetPointAsPartOfSegmentOnPlaneEarth(PointLatLng p1, PointLatLng p2, double part)
        {
            var lat = p1.Lat + (p2.Lat - p1.Lat) * part;
            var lng = p1.Lng + (p2.Lng - p1.Lng) * part;

            return new PointLatLng(lat, lng);
        }

        public static int GpsInSorFormat(double degrees)
        {
            int d = (int)degrees;
            double m = (degrees - d) * 60;
            int mi = (int)m;
            double s = (m - mi) * 60;
            int ss = (int)(s * 100);
            if (s * 100 - ss >= 0.5) ss++;
            return d * 1000000 + mi * 10000 + ss;
        }

        public static PointLatLng InTheMiddle(PointLatLng p1, PointLatLng p2)
        {
            return new PointLatLng((p1.Lat + p2.Lat) / 2, (p1.Lng + p2.Lng) / 2);
        }
    }
}

File: WebFt30/src/app/features/rtus/landmarks/gis-lab-calculator.ts
import * as L from 'leaflet';

enum DecartAxis {
  X,
  Y,
  Z
}

interface DecartCoors {
  X: number;
  Y: number;
  Z: number;
}

class CoorsInRad {
  Lat: number;
  Lng: number;

  // несколько конструкторов нельзя, поэтому такой гибрид
  constructor(p?: L.LatLng | number, lng?: number) {
    if (p instanceof L.LatLng) {
      // Конструктор от LatLng
      this.Lat = (p.lat * Math.PI) / 180;
      this.Lng = (p.lng * Math.PI) / 180;
    } else if (typeof p === 'number' && typeof lng === 'number') {
      // Конструктор от двух чисел
      this.Lat = p;
      this.Lng = lng;
    } else {
      // Пустой конструктор
      this.Lat = 0;
      this.Lng = 0;
    }
  }
}

const EarthRadius = 6372795;

export class GisCalc {
  getDistanceBetweenPointLatLng(p1: L.LatLng, p2: L.LatLng): number {
    const q1 = new CoorsInRad(p1);
    const q2 = new CoorsInRad(p2);
    return this.getDistanceBetweenPoints(q1, q2);
  }

  getDistanceBetweenPoints(q1: CoorsInRad, q2: CoorsInRad): number {
    let decCoors = this.sphereToDecart(q2);
    decCoors = this.rotateAround(decCoors, q1.Lng, DecartAxis.Z);
    decCoors = this.rotateAround(decCoors, Math.PI / 2 - q1.Lat, DecartAxis.Y);
    const q = this.decartToSphere(decCoors);
    const distanceOnSphere = Math.PI / 2 - q.Lat;
    return distanceOnSphere * EarthRadius;
  }

  sphereToDecart(sphere: CoorsInRad): DecartCoors {
    const p = Math.cos(sphere.Lat);
    return {
      Z: Math.sin(sphere.Lat),
      Y: p * Math.sin(sphere.Lng),
      X: p * Math.cos(sphere.Lng)
    };
  }

  decartToSphere(coors: DecartCoors): CoorsInRad {
    const hypot = Math.sqrt(coors.X * coors.X + coors.Y * coors.Y);
    return new CoorsInRad(Math.atan2(coors.Z, hypot), Math.atan2(coors.Y, coors.X));
  }

  rotateAround(coors: DecartCoors, alpha: number, axis: DecartAxis): DecartCoors {
    switch (axis) {
      case DecartAxis.X:
        return {
          X: coors.X,
          Y: coors.Y * Math.cos(alpha) + coors.Z * Math.sin(alpha),
          Z: -coors.Y * Math.sin(alpha) + coors.Z * Math.cos(alpha)
        };
      case DecartAxis.Y:
        return {
          X: -coors.Z * Math.sin(alpha) + coors.X * Math.cos(alpha),
          Y: coors.Y,
          Z: coors.Z * Math.cos(alpha) + coors.X * Math.sin(alpha)
        };
      default: // DecartAxis.Z
        return {
          X: coors.X * Math.cos(alpha) + coors.Y * Math.sin(alpha),
          Y: -coors.X * Math.sin(alpha) + coors.Y * Math.cos(alpha),
          Z: coors.Z
        };
    }
  }
}
